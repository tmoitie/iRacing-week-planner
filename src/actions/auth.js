import {
  getAuth, signOut as fbSignOut, signInWithEmailAndPassword, createUserWithEmailAndPassword,
  onAuthStateChanged, sendPasswordResetEmail, GoogleAuthProvider, signInWithPopup,
} from 'firebase/auth';
import { debouncedDispatcherSaveSettings, getSettingsFromFirebase, saveSettingsToFirebase } from './settings';

export const LOADING_AUTH = 'AUTH/LOADING_SIGN_IN';
export const ERROR_AUTH = 'AUTH/ERROR_AUTH';
export const ERROR_RESET = 'AUTH/ERROR_RESET';
export const ERROR_ACKNOWLEDGE = 'AUTH/ERROR_ACKNOWLEDGE';
export const SIGNED_IN = 'AUTH/SIGNED_IN';
export const SIGNED_OUT = 'AUTH/SIGNED_OUT';
export const RESET_SENT = 'AUTH/RESET_SENT';
export const LOADING_RESET = 'AUTH/LOADING_RESET';

export function signOut() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    if (getState().settings.firebaseSynced === false) {
      debouncedDispatcherSaveSettings.cancel();
      await dispatch(saveSettingsToFirebase());
    }
    const auth = getAuth(getState().auth.firebaseApp);
    await fbSignOut(auth);

    dispatch({ type: SIGNED_OUT });
  };
}

export function signIn(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await signInWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function signInWithGoogle() {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);

      const provider = new GoogleAuthProvider();
      provider.addScope('profile');
      provider.addScope('email');

      await signInWithPopup(auth, provider);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function createAccount(email, password) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_AUTH });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await createUserWithEmailAndPassword(auth, email, password);
      return {};
    } catch (error) {
      dispatch({ type: ERROR_AUTH, error });
      return { type: ERROR_AUTH };
    }
  };
}

export function acknowledgeAuthError() {
  return { type: ERROR_ACKNOWLEDGE };
}

export function signedIn(user) {
  return async (dispatch) => {
    await dispatch({ type: SIGNED_IN, user });

    if (user) {
      await dispatch(getSettingsFromFirebase());
    }
  };
}

export function startListener() {
  return async (dispatch, getState) => {
    const auth = getAuth(getState().auth.firebaseApp);
    const { currentUser } = auth;
    dispatch(signedIn(currentUser));

    onAuthStateChanged(auth, (user) => {
      dispatch(signedIn(user));
    });
  };
}

export function forgottenPassword(email) {
  return async (dispatch, getState) => {
    dispatch({ type: LOADING_RESET });
    try {
      const auth = getAuth(getState().auth.firebaseApp);
      await sendPasswordResetEmail(auth, email);
      dispatch({ type: RESET_SENT });
      return { type: RESET_SENT };
    } catch (error) {
      dispatch({ type: ERROR_RESET, error });
      return { type: ERROR_RESET };
    }
  };
}
