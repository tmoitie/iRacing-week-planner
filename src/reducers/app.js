// @flow

import moment, { duration } from 'moment';
import { CHANGE_MODAL, UPDATE_DAYS } from '../actions/app';
import { SIGNED_IN } from '../actions/auth';
import { seasonStart, weekSeasonStart, seasonEnd } from '../config';

const getWeek = (date) => Math.ceil(duration(moment(date).add({ second: 1 }).diff(weekSeasonStart)).asWeeks());

type AppState = {
  date: moment$Moment,
  daysSinceSeasonStart: number,
  week: number,
  currentModal: string,
};

type AppReducerArgs = {
  type?: string,
  days?: number,
  modalName?: string,
};

export default function app(
  initState: ?AppState,
  { type, days, modalName }: AppReducerArgs,
): AppState {
  let state = initState;

  if (state === undefined) {
    let currentDate = moment(new Date()).utc().startOf('day');

    if (currentDate.isBefore(seasonStart)) {
      currentDate = moment(seasonStart).utc().startOf('day');
    }

    if (currentDate.isAfter(seasonEnd)) {
      currentDate = moment(seasonEnd).utc().startOf('day').subtract({ days: 1 });
    }

    const currentDays = currentDate.diff(seasonStart, 'days');

    state = {
      date: currentDate,
      daysSinceSeasonStart: currentDays,
      week: getWeek(currentDate),
      currentModal: null,
    };
  }

  if (type === UPDATE_DAYS) {
    const date = moment(seasonStart).add(days, 'days');
    return {
      ...state,
      date,
      daysSinceSeasonStart: days,
      week: getWeek(date),
    };
  }

  if (type === CHANGE_MODAL) {
    return {
      ...state,
      currentModal: modalName,
    };
  }

  if (type === SIGNED_IN) {
    return {
      ...state,
      currentModal: null,
    };
  }

  return state;
}
