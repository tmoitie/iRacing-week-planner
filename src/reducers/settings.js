// @flow

import uniq from 'lodash.uniq';
import { actionTypes as localStorageActionTypes } from 'redux-localstorage';
import { SIGNED_OUT } from '../actions/auth';
import {
  FIREBASE_SYNCED,
  LOAD_SETTINGS_FROM_FIREBASE,
  RESET_FILTERS,
  RESET_SETTINGS,
  UPDATE_FILTERS,
  UPDATE_SETTING,
} from '../actions/settings';

import availableColumns from '../data/availableColumns';
import { tracks, cars } from '../data';

export type SortOptions = {
  key: string,
  order: string,
};

export type FilterOptions = {
  type: Array<string>,
  licence: Array<string>,
  official: Array<boolean>,
  fixed: boolean[],
  ownedCars: boolean,
  ownedTracks: boolean,
  favouriteSeries: boolean,
  favouriteCarsOnly: boolean,
  favouriteTracksOnly: boolean,
};

export const defaultFilters: FilterOptions = {
  type: ['Sports Car', 'Formula Car', 'Oval', 'Dirt', 'RX'],
  licence: ['R', 'D', 'C', 'B', 'A', 'P'],
  official: [false, true],
  fixed: [false, true],
  ownedCars: false,
  ownedTracks: false,
  favouriteSeries: false,
  favouriteTracksOnly: false,
  favouriteCarsOnly: false,
};

export type SettingOptions = {
  filters: FilterOptions,
  ownedCars: Array<number>,
  ownedTracks: Array<number>,
  favouriteSeries: Array<number>,
  favouriteCars: Array<number>,
  favouriteTracks: Array<number>,
  sort: SortOptions,
  columns: Array<number>,
  firebaseSynced: boolean,
};

export const defaultSettings: SettingOptions = {
  filters: defaultFilters,
  ownedCars: cars.filter((car) => car.freeWithSubscription === true).map((car) => car.sku),
  ownedTracks: tracks.filter((track) => track.default).map((track) => track.pkgid),
  favouriteSeries: [],
  favouriteCars: [],
  favouriteTracks: [],
  sort: { key: 'licence', order: 'asc' },
  columns: availableColumns.filter((column) => column.default === true).map((column) => column.id),
  firebaseSynced: false,
};

function synchroniseLegacyChanges(state) {
  const newState = { ...state };

  newState.ownedCars = uniq([
    ...newState.ownedCars,
    ...defaultSettings.ownedCars,
  ]);

  newState.ownedTracks = uniq([
    ...newState.ownedTracks,
    ...defaultSettings.ownedTracks,
  ]);

  const hasRoadFilter = newState.filters.type.includes('Road');

  if (hasRoadFilter) {
    newState.filters.type = [...newState.filters.type, 'Sports Car', 'Formula Car'];
    newState.filters.type = newState.filters.type.filter((t) => t !== 'Road');
  }

  return newState;
}

export default function settings(initState: SettingOptions, { type, payload }): SettingOptions {
  let state = initState;
  if (initState === undefined) {
    state = defaultSettings;
  }

  if (type === localStorageActionTypes.INIT) {
    if (!payload) {
      return state;
    }

    const newState = synchroniseLegacyChanges(payload.settings);

    return { ...state, ...newState };
  }

  if (type === UPDATE_FILTERS) {
    return { ...state, filters: payload.filters, firebaseSynced: false };
  }

  if (type === RESET_FILTERS) {
    return { ...state, filters: defaultFilters, firebaseSynced: false };
  }

  if (type === RESET_SETTINGS || type === SIGNED_OUT) {
    return { ...defaultSettings, firebaseSynced: false };
  }

  if (type === UPDATE_SETTING) {
    return { ...state, [payload.key]: payload.value, firebaseSynced: false };
  }

  if (type === LOAD_SETTINGS_FROM_FIREBASE) {
    return synchroniseLegacyChanges({
      ...state,
      ...defaultSettings,
      ...payload,
      firebaseSynced: true,
    });
  }

  if (type === FIREBASE_SYNCED) {
    return { ...state, firebaseSynced: true };
  }

  return state;
}
