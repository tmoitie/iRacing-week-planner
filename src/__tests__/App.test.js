import { afterEach, beforeEach, describe, test } from '@jest/globals';
import moment from 'moment';
import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import MockDate from 'mockdate';
import * as firebaseAuth from 'firebase/auth';
import { SIGNED_IN } from '../actions/auth';

import App from '../App';

import '../data/season.json';
import { defaultFilters } from '../reducers/settings';

jest.mock('firebase/auth');
jest.mock('../data/season.json');

const mockStore = configureMockStore([thunk]);

describe('components/App', () => {
  const defaultStore = {
    settings: {
      sort: { key: 'id', order: 'asc' },
      filters: defaultFilters,
      favouriteSeries: [],
      ownedTracks: [],
      ownedCars: [],
      favouriteCars: [],
      favouriteTracks: [],
      columns: ['id'],
    },
    app: {
      date: moment('2022-09-10T00:00:00.000Z'),
      daysSinceSeasonStart: 4,
      week: 1,
      currentModal: null,
    },
    auth: {
      user: null,
      firebaseApp: {},
    },
  };

  beforeEach(() => {
    MockDate.set('2022-09-07T13:30:00.000Z');
  });

  afterEach(() => {
    MockDate.reset();
  });

  test('renders correctly', () => {
    const store = mockStore(defaultStore);
    const { container } = render(<Provider store={store}><App /></Provider>);

    expect(container.firstChild).toMatchSnapshot();
    expect(store.getActions()[0].type).toEqual(SIGNED_IN);
    expect(store.getActions()[0].user).not.toBeDefined();

    const newUser = { id: 123 };
    firebaseAuth.testDispatchOnAuthStateChanged(newUser);
    expect(store.getActions()[1].type).toEqual(SIGNED_IN);
    expect(store.getActions()[1].user).toBe(newUser);
  });
});
