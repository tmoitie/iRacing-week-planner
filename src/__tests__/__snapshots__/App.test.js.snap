// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/App renders correctly 1`] = `
<div>
  <nav
    class="navbar navbar-inverse"
  >
    <div
      class="container-fluid"
    >
      <div
        class="navbar-header"
      >
        <a
          class="navbar-brand"
          href="/"
        >
          iRacing Week Planner
        </a>
      </div>
      <ul
        class="nav navbar-nav navbar-left"
      >
        <li>
          <a
            href=""
            id="navbar-link-purchase-guide"
          >
            Purchase guide
          </a>
        </li>
      </ul>
      <ul
        class="nav navbar-nav navbar-right"
      >
        <li>
          <a
            href=""
            id="navbar-link-my-tracks"
          >
            Set my tracks
          </a>
        </li>
        <li>
          <a
            href=""
            id="navbar-link-my-cars"
          >
            Set my cars
          </a>
        </li>
        <li>
          <a
            href=""
            id="navbar-link-favourite-series"
          >
            Set favorite series
          </a>
        </li>
        <li>
          <a
            href=""
            id="navbar-link-options"
          >
            Options
          </a>
        </li>
        <li>
          <a
            href=""
            id="navbar-link-about"
          >
            About
          </a>
        </li>
        <li>
          <a
            href=""
            id="navbar-link-sign-in"
          >
            Sign in
          </a>
        </li>
        <li
          class="dropdown"
          id="navbar-language-dropdown"
        >
          <a
            href=""
            id="navbar-link-open-language-dropdown"
          >
            ðŸ‡ºðŸ‡¸
             
            <span
              class="caret"
            />
          </a>
          <ul
            class="dropdown-menu"
          >
            <li>
              <a
                href=""
                id="navbar-link-language-de"
              >
                ðŸ‡©ðŸ‡ª
                 
                Deutsch (DE)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-en"
              >
                ðŸ‡ºðŸ‡¸
                 
                English (US)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-en-GB"
              >
                ðŸ‡¬ðŸ‡§
                 
                English (UK)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-es"
              >
                ðŸ‡ªðŸ‡¸
                 
                EspaÃ±ol (ES)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-fr"
              >
                ðŸ‡«ðŸ‡·
                 
                FranÃ§ais (FR)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-it"
              >
                ðŸ‡®ðŸ‡¹
                 
                Italiano (IT)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-ja"
              >
                ðŸ‡¯ðŸ‡µ
                 
                æ—¥æœ¬èªž (JA)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-nl"
              >
                ðŸ‡³ðŸ‡±
                 
                Nederlands (NL)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-pl"
              >
                ðŸ‡µðŸ‡±
                 
                Polski (PL)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-pt-BR"
              >
                ðŸ‡§ðŸ‡·
                 
                PortuguÃªs (BR)
              </a>
            </li>
            <li>
              <a
                href=""
                id="navbar-link-language-tr"
              >
                ðŸ‡¹ðŸ‡·
                 
                TÃ¼rkÃ§e (TR)
              </a>
            </li>
            <li>
              <a
                href="https://github.com/tmoitie/iRacing-week-planner/blob/master/Translate.md"
                rel="noreferrer"
                target="_blank"
              >
                Help me translate!
              </a>
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </nav>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-md-2"
      >
        <div>
          <button
            class="bmcButton"
            type="button"
          >
            <img
              alt="Buy me a coffee"
              src="https://cdn.buymeacoffee.com/buttons/bmc-new-btn-logo.svg"
            />
            <span
              style="margin-left: 5px;"
            >
              Buy me a coffee
            </span>
          </button>
          <div
            class="sponsoredContainer"
          >
            <a
              class="sponsoredLink"
              href="https://coachdaveacademy.com/product/iracing-setup-subscription/?ref=weekly_planner"
              rel="sponsored"
            >
              <h4
                class="sponsoredHeader"
              >
                Sponsored
              </h4>
              <h3
                class="sponsoredTitle"
              >
                Looking for iRacing Setups?
              </h3>
              <p
                class="sponsoredBody"
              >
                Grab a setup from Coach Dave Academy, developed by professional sim racers and engineers.
              </p>
              <p
                class="btn btn-primary button"
              >
                Browse Setups
              </p>
            </a>
          </div>
        </div>
        <h3>
          Filters
        </h3>
        <div
          style="font-size: 0.8em;"
        >
          <h4>
            Type
          </h4>
          <div
            class="checkbox"
            data-testid="checkbox-type-oval"
          >
            <label
              for="checkbox-type-oval"
            >
              <input
                checked=""
                id="checkbox-type-oval"
                type="checkbox"
              />
              Oval
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-type-road-sports"
          >
            <label
              for="checkbox-type-road-sports"
            >
              <input
                checked=""
                id="checkbox-type-road-sports"
                type="checkbox"
              />
              Sports Car
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-type-road-formula"
          >
            <label
              for="checkbox-type-road-formula"
            >
              <input
                checked=""
                id="checkbox-type-road-formula"
                type="checkbox"
              />
              Formula Car
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-type-dirt"
          >
            <label
              for="checkbox-type-dirt"
            >
              <input
                checked=""
                id="checkbox-type-dirt"
                type="checkbox"
              />
              Dirt
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-type-rx"
          >
            <label
              for="checkbox-type-rx"
            >
              <input
                checked=""
                id="checkbox-type-rx"
                type="checkbox"
              />
              RX
            </label>
          </div>
          <h4>
            Licence
          </h4>
          <div
            class="checkbox"
            data-testid="checkbox-licence-r"
          >
            <label
              for="checkbox-licence-r"
            >
              <input
                checked=""
                id="checkbox-licence-r"
                type="checkbox"
              />
              R
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-licence-d"
          >
            <label
              for="checkbox-licence-d"
            >
              <input
                checked=""
                id="checkbox-licence-d"
                type="checkbox"
              />
              D
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-licence-c"
          >
            <label
              for="checkbox-licence-c"
            >
              <input
                checked=""
                id="checkbox-licence-c"
                type="checkbox"
              />
              C
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-licence-b"
          >
            <label
              for="checkbox-licence-b"
            >
              <input
                checked=""
                id="checkbox-licence-b"
                type="checkbox"
              />
              B
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-licence-a"
          >
            <label
              for="checkbox-licence-a"
            >
              <input
                checked=""
                id="checkbox-licence-a"
                type="checkbox"
              />
              A
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-licence-p"
          >
            <label
              for="checkbox-licence-p"
            >
              <input
                checked=""
                id="checkbox-licence-p"
                type="checkbox"
              />
              P
            </label>
          </div>
          <h4>
            Official/Fixed
          </h4>
          <div
            class="checkbox"
            data-testid="checkbox-official-false"
          >
            <label
              for="checkbox-official-false"
            >
              <input
                checked=""
                id="checkbox-official-false"
                type="checkbox"
              />
              Unofficial
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-official-true"
          >
            <label
              for="checkbox-official-true"
            >
              <input
                checked=""
                id="checkbox-official-true"
                type="checkbox"
              />
              Official
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-fixed-false"
          >
            <label
              for="checkbox-fixed-false"
            >
              <input
                checked=""
                id="checkbox-fixed-false"
                type="checkbox"
              />
              Open setup
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-fixed-true"
          >
            <label
              for="checkbox-fixed-true"
            >
              <input
                checked=""
                id="checkbox-fixed-true"
                type="checkbox"
              />
              Fixed setup
            </label>
          </div>
          <h4>
            Content
          </h4>
          <div
            class="checkbox"
            data-testid="checkbox-ownedCars"
          >
            <label
              for="checkbox-ownedCars"
            >
              <input
                id="checkbox-ownedCars"
                type="checkbox"
              />
              Owned cars only
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-ownedTracks"
          >
            <label
              for="checkbox-ownedTracks"
            >
              <input
                id="checkbox-ownedTracks"
                type="checkbox"
              />
              Owned tracks only
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-favouriteSeries"
          >
            <label
              for="checkbox-favouriteSeries"
            >
              <input
                id="checkbox-favouriteSeries"
                type="checkbox"
              />
              Favorite series only
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-favouriteCarsOnly"
          >
            <label
              for="checkbox-favouriteCarsOnly"
            >
              <input
                id="checkbox-favouriteCarsOnly"
                type="checkbox"
              />
              Favorite cars only
            </label>
          </div>
          <div
            class="checkbox"
            data-testid="checkbox-favouriteTracksOnly"
          >
            <label
              for="checkbox-favouriteTracksOnly"
            >
              <input
                id="checkbox-favouriteTracksOnly"
                type="checkbox"
              />
              Favorite tracks only
            </label>
          </div>
          <p>
            <button
              class="btn btn-primary"
              id="filters-reset-filters-button"
              type="button"
            >
              Reset filters
            </button>
          </p>
          <p>
            <button
              class="btn btn-primary"
              id="filters-reset-settings-button"
              type="button"
            >
              Reset all settings
            </button>
          </p>
        </div>
      </div>
      <div
        class="col-md-10"
      >
        <div
          class="row"
        >
          <h3
            class="col-xs-8"
          >
            Races for date: September 10, 2022
          </h3>
          <h3
            class="col-xs-4"
            style="text-align: right;"
          >
            Week 1
          </h3>
        </div>
        <div
          style="margin-bottom: 10px;"
        >
          <div
            class="bp5-slider bp5-slider-unlabeled"
          >
            <div
              class="bp5-slider-track"
            >
              <div
                class="bp5-slider-progress"
                style="left: 0.00%; right: 100.00%; top: 0px;"
              />
              <div
                class="bp5-slider-progress bp5-intent-primary"
                style="left: 0.00%; right: 95.60%; top: 0px;"
              />
              <div
                class="bp5-slider-progress"
                style="left: 4.40%; right: 0.00%; top: 0px;"
              />
            </div>
            <div
              class="bp5-slider-axis"
            />
            <span
              aria-disabled="false"
              aria-orientation="horizontal"
              aria-valuemax="91"
              aria-valuemin="0"
              aria-valuenow="4"
              class="bp5-slider-handle"
              role="slider"
              style="left: calc(4.40% - 0px);"
              tabindex="0"
            />
          </div>
        </div>
        <div
          class="table-responsive raceListing"
        >
          <table
            class="table"
            style="font-size: 0.8em;"
          >
            <thead>
              <tr>
                <th
                  class="clickableCell"
                  id="raceListing-th-id"
                >
                  <button
                    class="cellButton"
                    type="button"
                  >
                    ID
                    <span>
                       
                    </span>
                    <span
                      class="glyphicon glyphicon-triangle-top"
                    />
                  </button>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>
                  <div>
                    245
                  </div>
                </td>
              </tr>
              <tr>
                <td>
                  <div>
                    374
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
